### PR TITLE
Feature/stp basic implementation

### PR DESCRIPTION
**Pull request by timovdk**
Basic implementation of the new STP model. For this PR to work, make sure you have the latest roboteam_utils. #844

### CODE REVIEW

File: test/StpTests/PlayCheckerTests.cpp [hunk 1]
```diff
@@ -15,6 +15,10 @@ class AlwaysValid : public rtt::ai::stp::Play {
     uint8_t score(rtt::world_new::World *world) noexcept override {
         return 100;
     }
+
+    void assignRoles() noexcept override {
+
```
Comment (timovdk) [inline test/StpTests/PlayCheckerTests.cpp]:
```suggestion
```

File: test/StpTests/PlayCheckerTests.cpp [hunk 2]
```diff
@@ -25,6 +29,10 @@ class AlwaysFalse : public rtt::ai::stp::Play {
     uint8_t score(rtt::world_new::World *world) noexcept override {
         return 0;
     }
+
+    void assignRoles() noexcept override {
+
```
Comment (timovdk) [inline test/StpTests/PlayCheckerTests.cpp]:
```suggestion
```

File: test/StpTests/PlayDeciderTests.cpp [hunk 3]
```diff
@@ -16,6 +16,10 @@ class AlwaysValid : public rtt::ai::stp::Play {
     uint8_t score(rtt::world_new::World *world) noexcept override {
         return 100;
     }
+
+    void assignRoles() noexcept override {
+
```
Comment (timovdk) [inline test/StpTests/PlayDeciderTests.cpp]:
```suggestion
```

File: test/StpTests/PlayDeciderTests.cpp [hunk 4]
```diff
@@ -26,6 +30,10 @@ class AlwaysFalse : public rtt::ai::stp::Play {
     uint8_t score(rtt::world_new::World *world) noexcept override {
         return 0;
     }
+
+    void assignRoles() noexcept override {
+
```
Comment (timovdk) [inline test/StpTests/PlayDeciderTests.cpp]:
```suggestion
```

File: src/stp/new_tactics/TestTactic.cpp [hunk 5]
```diff
@@ -2,36 +2,45 @@
...
+    skills.initialize();
+}
+
 void TestTactic::onInitialize() noexcept {
-    skills = rtt::collections::state_machine<Skill, Status, SkillInfo>{Rotate(), SetDribbler(), Rotate()};
 
-    // Initialize first skill
-    skills.initialize();
+
+
```
Comment (timovdk) [inline src/stp/new_tactics/TestTactic.cpp]:
```suggestion
```

File: src/stp/Play.cpp [hunk 6]
```diff
@@ -5,27 +5,50 @@
...
+        RTT_WARNING("Reassigning bots");
+        assignRoles();
+    }
 
-        if (count[static_cast<size_t>(Status::Waiting)]) {
-            return Status::Waiting;
+    for (auto& each : roles) {
+        auto roleName{each->getName()};
+        if(tacticInfos.find(roleName) != tacticInfos.end()) {
+
```
Comment (timovdk) [inline src/stp/Play.cpp]:
```suggestion
```

File: src/stp/new_tactics/TestTactic.cpp [hunk 7]
```diff
@@ -2,36 +2,43 @@
...
 namespace rtt::ai::stp {
 
+TestTactic::TestTactic(){
+    // Create state machine of skills and initialize first skill
+    skills = rtt::collections::state_machine<Skill, Status, SkillInfo>{GoToPos(), Rotate()};
+    skills.initialize();
+}
+
 void TestTactic::onInitialize() noexcept {
-    skills = rtt::collections::state_machine<Skill, Status, SkillInfo>{Rotate(), SetDribbler(), Rotate()};
```
Comment (timovdk) [inline src/stp/new_tactics/TestTactic.cpp]:
```suggestion
```

File: include/roboteam_ai/stp/Play.hpp [hunk 8]
```diff
@@ -5,62 +5,106 @@
...
 
 namespace rtt::ai::stp {
 
+/**
+ * Play class that's used in the STP model
+ * on update traverses every Role, and updates it.
+ */
+class Play {
+   public:
+    constexpr static size_t ROBOT_COUNT = 11;
```
Comment (Yuhanun) [inline include/roboteam_ai/stp/Play.hpp]: I know I wrote this code but I kinda beuned it, this should be in constants.h tbh

File: include/roboteam_ai/utilities/Dealer.h [hunk 9]
```diff
@@ -50,6 +50,8 @@ class Dealer {
   using FlagMap = std::map<std::string, std::vector<DealerFlag>>;
   Dealer(v::WorldDataView world, world::Field * field);
   virtual ~Dealer() = default; // needed for test
+  // Create a distribution of robots
+  std::unordered_map<std::string, v::RobotView> distribute(const std::vector<v::RobotView> &allRobots, const FlagMap &flagMap);
```
Comment (Yuhanun) [inline include/roboteam_ai/utilities/Dealer.h:54]: why std::string

**Review by Yuhanun — APPROVED** [review]
LGTM

File: include/roboteam_ai/stp/Play.hpp (hunk 8 above)
Comment (timovdk) [reply include/roboteam_ai/stp/Play.hpp]: Done

File: include/roboteam_ai/utilities/Dealer.h (hunk 9 above)
Comment (timovdk) [reply include/roboteam_ai/utilities/Dealer.h:54]: This is the role_name the robot is distributed to. We might be able to make some better solution for this but that is a refactor for a later PR if necessary

**Pull request merged**